### PR TITLE
fix(TX-824): Type guard for credit card picker in new payment

### PR DESCRIPTION
**Pull request by AlicanAkyuz**
The type of this PR is: **Fix**

This PR solves [TX-824]

### Description
This PR attempts to fix the problem that buyers cannot fix a failed payment with credit card initially and have to try a second time to succeed. 

The detected problem was that by the time the buyer is prompted to authenticate their purchase through 3D Security, our credit card picker has already presented the creditCardId. When the buyer re-lands on the page after authorising the purchase through the Stripe modal, we lose the creditCardId initially acquired and make an API call to MP's newPayment mutation with creditCardId missing in the payload. 

This PR makes sure that we have the creditCardId after the 3D security modal has been closed. 

#### Video -> Make offer, accept offer, payment fails, user sets new payment with 3D
https://user-images.githubusercontent.com/42584148/199185863-f74c74f0-c93c-4e89-b45c-fa395a4a11a1.mov

[TX-824]: https://artsyproduct.atlassian.net/browse/TX-824?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Apps/Order/Routes/NewPayment/index.tsx [hunk 1]
```diff
@@ -64,62 +75,80 @@ export const NewPaymentRoute: FC<NewPaymentProps & StripeProps> = props => {
...
+    if (result?.type === "internal_error") {
+      dialog.showErrorDialog({
+        title: "An internal error occurred",
+      })
+      logger.error(result?.error)
+      return null
+    }
+
+    if (result?.type === "success" && result?.creditCardId) {
+      setCreditCardPickerResult(result)
```
Comment (starsirius) [inline src/Apps/Order/Routes/NewPayment/index.tsx:105]: 👍

**Review by starsirius — APPROVED** [review]
Have a non-blocking comment to make the `CreditCardPicker` component be responsible for maintaining credit card selection state; otherwise, looks great to me! Thanks for the quick fix!

Related to the checkout improvement epic, curious your thoughts on tests to cover various scenarios here.

File: src/Apps/Order/Routes/NewPayment/index.tsx [hunk 2]
```diff
@@ -47,10 +47,21 @@ export interface NewPaymentProps {
...
+  | { type: "success"; creditCardId: string }
+
 const logger = createLogger("Order/Routes/NewPayment/index.tsx")
 
 export const NewPaymentRoute: FC<NewPaymentProps & StripeProps> = props => {
   const [isGettingCreditCardId, setIsGettingCreditCardId] = useState(false)
+  const [
+    creditCardPickerResult,
+    setCreditCardPickerResult,
+  ] = useState<CreditCardPickerResultType | null>(null)
```
[the remark this answers is not in the document]
Comment (AlicanAkyuz) [reply src/Apps/Order/Routes/NewPayment/index.tsx:63]: thank you for this suggestion. I agree that it'd read more straightforward to have that logic in the CreditCard picker. I'll ship this fix and consider that option within the scope of the checkout flow improvements (we'll work on the new payment page in more detail, hopefully soon 🤞 )

**Comment by AlicanAkyuz** [comment]
@starsirius thank you! to be honest, I didn't even know that we have this kind of 3D security check in this flow. I have currently no idea how to mock and test that scenario, but I'll research on it to see what we can do moving forward.

**Pull request merged**